core/state/state_object.go

```diff
@@ -228,15 +221,11 @@ func (s *stateObject) SetState(key, value common.Hash) {
 		prevvalue = &prev
 	}
 	// New value is different, update and journal the change
-	s.db.journal.append(storageChange{
-		account:   &s.address,
-		key:       key,
-		prevvalue: prevvalue,
-	})
+	s.db.journal.JournalSetState(s.address, key, prevvalue)
+	s.setState(key, &value)
```

Comment:
The `logger` invocation is moved down, yes, because it's not "core". I wanted it to not be mixed in with the actual consensus code. 

Since the values are already loaded into variables, it should be fine to do so.

Reply:
But every single other call has the logger happening first and then the state change.

Reply:
Yes, because consider this one: 

```golang
    if s.db.logger != nil && s.db.logger.OnBalanceChange != nil {
        s.db.logger.OnBalanceChange(s.address, s.Balance().ToBig(), amount.ToBig(), reason)
   }
```
In the call to the `logger`, it actually uses the `s.Balance()`. So it needs to be done before setting the new value. Alternatively, stash the value in a variable, so the `logger`-call later on can use that.

So we're left with two crappy choices: introduce a (possibly unused) tempvar, or or do the logging-call in the middle of the core consensus mechanics.